PhysicsTools/PatUtils/plugins/PackedCandidatesTrackLiteModifier.cc

```diff
@@ -0,0 +1,72 @@
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+#include <vector>
+
+class PackedCandidatesTrackLiteModifier : public edm::global::EDProducer<> {
+public:
+  explicit PackedCandidatesTrackLiteModifier(const edm::ParameterSet &);
+  ~PackedCandidatesTrackLiteModifier() override = default;
+
+  void produce(edm::StreamID, edm::Event &, const edm::EventSetup &) const override;
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+
+private:
+  const edm::EDGetTokenT<pat::PackedCandidateCollection> inputCandidates_;
+  const edm::EDPutTokenT<pat::PackedCandidateCollection> outputCandidates_;
+  const unsigned int covSchema_;
+  const unsigned int covVersion_;
+  const unsigned int nHits_;
+  const unsigned int nPixelHits_;
+};
+
+//____________________________________________________________________________||
+PackedCandidatesTrackLiteModifier::PackedCandidatesTrackLiteModifier(const edm::ParameterSet &iConfig)
+    : inputCandidates_(
+          consumes<pat::PackedCandidateCollection>(iConfig.getParameter<edm::InputTag>("inputCandidates"))),
+      outputCandidates_(produces<pat::PackedCandidateCollection>("")),
+      covSchema_(iConfig.getParameter<unsigned int>("covSchema")),
+      covVersion_(iConfig.getParameter<unsigned int>("covVersion")),
+      nHits_(iConfig.getParameter<unsigned int>("nHits")),
+      nPixelHits_(iConfig.getParameter<unsigned int>("nPixelHits")) {
+}
+
+void PackedCandidatesTrackLiteModifier::produce(edm::StreamID, edm::Event &iEvent, const edm::EventSetup &iSetup) const {
+
+  auto const& packedCandidates = iEvent.get(inputCandidates_);
+
+  pat::PackedCandidateCollection output;
+  output.reserve(packedCandidates.size());
+  for (auto const& cand: packedCandidates) {
+    output.push_back(pat::PackedCandidate(cand));
+
+    if (!output.back().hasTrackDetails() && output.back().covarianceVersion() == int(covVersion_))
```

Comment:
there was a discussion earlier on (im)possibility to use `covarianceVersion == 0` in this context.
Have you looked into this?

The implementation of interest (UL miniAOD) of PATPackedCandidateProducer 
```C++
    } else {
      if (!PVs->empty()) {
        PV = reco::VertexRef(PVs, 0);
        PVpos = PV->position();
      }

      outPtrP->push_back(pat::PackedCandidate(
          cand.polarP4(), PVpos, cand.pt(), cand.eta(), cand.phi(), cand.pdgId(), PVRefProd, PV.key()));
      outPtrP->back().setAssociationQuality(
          pat::PackedCandidate::PVAssociationQuality(pat::PackedCandidate::UsedInFitTight));
```
I can see a few alternatives for `covarianceVersion() == 0`:
- a simple check for a charged candidate with `UsedInFitTight` and without `hasTrackDetails`
- also check that the candidate `dphi_, deta_, dtrkpt_` are at zero (also, IIUC the `dxy_` and `dz_` are zero for these candidates as well)

What do you think?

Reply:
We ran some tests and the conclusions are
* there are candidates with covarianceVersion == 0 that are charged, which is why we did not want to rely only on charge
* dtrkpt_ and deta_ seem to almost always be 0, at least if computed from pt() - ptTrk() or eta() - etaAtVtx() (probably because there is a cutoff at 0.001 for both of them)
* phi() - phiAtVtx() is never 0, even in candidate with covarianceVersion == 0, possibly due to rounding while limimiting the range to [-Pi,Pi]
* dxy() and dzAssociatedPV(), which are dxy_ and dz_, seem however to be good indicators

The idea we are implementing is to remove the covarianceVersion check and instead check the charge, dxy() and dzAssociatedPV() of the candidate, a commit will come soon

Reply:
> also check that the candidate `dphi_, deta_, dtrkpt_` are at zero (also, IIUC the `dxy_` and `dz_` are zero for these candidates as well)

perhaps checking the packed values is even more straighforward; this apparently implies an extra test accessor is needed in the `PackedCandidate` class.

Reply:
We have added the `fromTrackCandidate()` (probably there's a better naming) method to check `packedDz_!=0 | | packedDxy_!=0` to distinguish between candidates that originally had no track.